waiter/src/waiter/auth/oidc.clj

```diff
@@ -0,0 +1,211 @@
+;;
+;; Copyright (c) Two Sigma Open Source, LLC
+;;
+;; Licensed under the Apache License, Version 2.0 (the "License");
+;; you may not use this file except in compliance with the License.
+;; You may obtain a copy of the License at
+;;
+;;  http://www.apache.org/licenses/LICENSE-2.0
+;;
+;; Unless required by applicable law or agreed to in writing, software
+;; distributed under the License is distributed on an "AS IS" BASIS,
+;; WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+;; See the License for the specific language governing permissions and
+;; limitations under the License.
+;;
+(ns waiter.auth.oidc
+  (:require [clojure.core.async :as async]
+            [clojure.string :as str]
+            [clojure.tools.logging :as log]
+            [waiter.auth.authentication :as auth]
+            [waiter.auth.jwt :as jwt]
+            [waiter.cookie-support :as cookie-support]
+            [waiter.status-codes :refer :all]
+            [waiter.util.http-utils :as hu]
+            [waiter.util.ring-utils :as ru]
+            [waiter.util.utils :as utils])
+  (:import (java.net URI)))
+
+(def ^:const accept-redirect-header-name "accept-redirect")
+
+(def ^:const accept-redirect-auth-header-name "accept-redirect-auth")
+
+(def ^:const challenge-cookie-duration-secs 60)
+
+(def ^:const code-verifier-length 128)
+
+(def ^:const content-security-policy-value "default-src 'none'; frame-ancestors 'none'")
+
+(def ^:const oidc-challenge-cookie "x-waiter-oidc-challenge")
+
+(def ^:const oidc-callback-uri "/oidc/v1/callback")
+
+(defn create-code-verifier
+  "Creates a randomly generated string containing characters from A-Z of length code-verifier-length."
+  []
+  (apply str (take code-verifier-length (repeatedly #(char (+ (rand 26) 65))))))
+
+(defn create-state-code
+  "Creates a Base64 encoded string of the input state map."
+  [password state-map]
+  (utils/map->base-36-string state-map password))
```

Comment:
To have the codes be properly url encoded, the presence of spaces was not being handled well after the redirect back to Waiter OIDC callback.

Reply:
I still don't follow.  Can you give me an example of the problem?

Reply:
This is an example using base64 encoding:

state code sent in authorize redirect: `TlBZAgpxgARfxpysL+8esoqcsJv4Fx6Z8Wqyzt2PKfAXNigL1b9QI0353JCU3pVJ7nHMmEpMhGrQ/RyHdKyjb8BRcK319IxkLQvvX7XlKi7crdwCuBiWKu/8kNN745qMITFTPw==`
state code received in OIDC callback: `TlBZAgpxgARfxpysL 8esoqcsJv4Fx6Z8Wqyzt2PKfAXNigL1b9QI0353JCU3pVJ7nHMmEpMhGrQ/RyHdKyjb8BRcK319IxkLQvvX7XlKi7crdwCuBiWKu/8kNN745qMITFTPw==`
Note the `+` being replaced by whitespaces.
Attempting to correct with URL encoding: `TlBZAgpxgARfxpysL+8esoqcsJv4Fx6Z8Wqyzt2PKfAXNigL1b9QI0353JCU3pVJ7nHMmEpMhGrQ%2FRyHdKyjb8BRcK319IxkLQvvX7XlKi7crdwCuBiWKu%2F8kNN745qMITFTPw%3D%3D`
ends up being broken since the last `=`s are replaced also.

Reply:
The conversion of `+` to ` ` is happening inside the call to `ru/query-params-request`:
```
{:strs [code state]} (-> request ru/query-params-request :query-params)
```

```
(URLDecoder/decode "a=b1+b2&c=d==" "utf-8")
=> "a=b1 b2&c=d=="
``` 
https://github.com/ring-clojure/ring-codec/blob/58cccc597668cef8a3840ac54c7b6b32d2b95990/src/ring/util/codec.clj#L123-L131

Reply:
Are the trailing '='s padding?

Reply:
Yes, generated from the base64 encoding